### PR TITLE
evm: params

### LINKED ISSUE #443: Custom Genesis Parameters
__Proposal:__ [Description of the feature]
A way for a genesis file with a more robust set of parameters to take in and generate a genesis state from.
- EIPs
- hardforks (DAO, Constantinople, Petersburg, etc.)
- other important genesis parameters

__Current behavior:__ [What currently happens]
only allows for tendermint related modules to be added to the genesis 
- this includes the bank module that will allow pre-funding at genesis

__Desired behavior:__ [What you would like to happen]
have the same state that can be emulated on an ethereum client (such as geth) to spin up the proper genesis state 

__Use case:__ [Why is this important (helps with prioritizing requests)]
this will be important if users would like to conform their network to the same as ethereum's mainnet.

short term: required to do a more close comparison of ethermint and geth.

Requests may be closed if we're not actively planning to work on them.

### PR DESCRIPTION
**Pull request by fedekunze**
Closes: #443 

## Description

TODO:

- [x] param tests
______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: x/evm/types/chain_config.go [hunk 1]
```diff
@@ -2,25 +2,171 @@ package types
...
 		DAOForkSupport:      true,
-		EIP150Block:         big.NewInt(0),
-		EIP150Hash:          common.HexToHash("0x2086799aeebeae135c246c65021c82b4e15a2c451340993aacfd2751886514f0"),
-		EIP155Block:         big.NewInt(0),
-		EIP158Block:         big.NewInt(0),
-		ByzantiumBlock:      big.NewInt(0),
-		ConstantinopleBlock: big.NewInt(0),
-		PetersburgBlock:     big.NewInt(0),
+		EIP150Block:         sdk.ZeroInt(),
+		EIP150Hash:          "0x2086799aeebeae135c246c65021c82b4e15a2c451340993aacfd2751886514f0",
```
Comment (fedekunze) [inline x/evm/types/chain_config.go]: I actually think this default hash is invalid should be populated on initialization

Comment (fedekunze) [reply x/evm/types/chain_config.go]: @noot @araskachoi thoughts?

**Comment by fedekunze** [comment]
the simulations are failing because the evm module is not supported yet on sims :

```
--- FAIL: TestAppSimulationAfterImport (101.00s)
panic: UnmarshalJSON cannot decode empty bytes [recovered]
	panic: UnmarshalJSON cannot decode empty bytes

goroutine 61 [running]:
testing.tRunner.func1.1(0x4d37e00, 0xc004197540)
	/usr/local/Cellar/go/1.15/libexec/src/testing/testing.go:1057 +0x30d
testing.tRunner.func1(0xc000602c00)
	/usr/local/Cellar/go/1.15/libexec/src/testing/testing.go:1060 +0x41a
panic(0x4d37e00, 0xc004197540)
	/usr/local/Cellar/go/1.15/libexec/src/runtime/panic.go:969 +0x175
github.com/cosmos/cosmos-sdk/x/params/subspace.Subspace.Get(0xc000266000, 0x51ae9e0, 0xc000023e50, 0x51aea20, 0xc000023ea0, 0xc000038190, 0x3, 0xb, 0xc0002038f0, 0x51bea60, ...)
	/Users/federico/go/pkg/mod/github.com/cosmos/cosmos-sdk@v0.39.1/x/params/subspace/subspace.go:106 +0x2ca
github.com/cosmos/cosmos-sdk/x/params/subspace.Subspace.GetParamSet(0xc000266000, 0x51ae9e0, 0xc000023e50, 0x51aea20, 0xc000023ea0, 0xc000038190, 0x3, 0xb, 0xc0002038f0, 0x51bea60, ...)
	/Users/federico/go/pkg/mod/github.com/cosmos/cosmos-sdk@v0.39.1/x/params/subspace/subspace.go:217 +0x1a8
github.com/cosmos/ethermint/x/evm/types.(*CommitStateDB).GetParams(...)
	/Users/federico/ethermint/x/evm/types/statedb.go:250
github.com/cosmos/ethermint/x/evm/keeper.Keeper.GetParams(0xc000266000, 0x51ae9e0, 0xc000023e80, 0xc00014f800, 0x0, 0xc0002ccd40, 0x51bea60, 0xc000126008, 0x51cc000, 0xc0021db280, ...)
	/Users/federico/ethermint/x/evm/keeper/params.go:11 +0x1b4
github.com/cosmos/ethermint/x/evm.ExportGenesis(0x51bea60, 0xc000126008, 0x51cc000, 0xc0021db280, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/federico/ethermint/x/evm/genesis.go:86 +0x279
github.com/cosmos/ethermint/x/evm.AppModule.ExportGenesis(0xc000266000, 0x51ae9e0, 0xc000023e80, 0xc00014f800, 0x0, 0xc0002ccd40, 0x51c20e0, 0xc000266a80, 0x51bea60, 0xc000126008, ...)
	/Users/federico/ethermint/x/evm/module.go:133 +0xa8
github.com/cosmos/cosmos-sdk/types/module.(*Manager).ExportGenesis(0xc000266b60, 0x51bea60, 0xc000126008, 0x51cc000, 0xc0021db280, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/federico/go/pkg/mod/github.com/cosmos/cosmos-sdk@v0.39.1/types/module/module.go:285 +0xfe
github.com/cosmos/ethermint/app.(*EthermintApp).ExportAppStateAndValidators(0xc000180b00, 0xc000122001, 0xc003119c70, 0x0, 0x0, 0x0, 0x0, 0x15, 0x0, 0x0, ...)
	/Users/federico/ethermint/app/export.go:40 +0x125
github.com/cosmos/ethermint/app.TestAppSimulationAfterImport(0xc000602c00)
	/Users/federico/ethermint/app/simulation_test.go:212 +0x71a
testing.tRunner(0xc000602c00, 0x50574d0)
	/usr/local/Cellar/go/1.15/libexec/src/testing/testing.go:1108 +0xef
created by testing.(*T).Run
	/usr/local/Cellar/go/1.15/libexec/src/testing/testing.go:1159 +0x386
FAIL	github.com/cosmos/ethermint/app	101.773s
```

**Comment by fedekunze** [comment]
Not sure what's going on with the storage rpc test tho:

```
=== RUN   TestEth_ExportAccount_WithStorage
    rpc_test.go:837: 
        	Error Trace:	rpc_test.go:837
        	Error:      	Not equal: 
        	            	expected: hexutil.Bytes{0x60, 0x80, 0x60, 0x40, 0x52, 0x34, 0x80, 0x15, 0x60, 0xf, 0x57, 0x60, 0x0, 0x80, 0xfd, 0x5b, 0x50, 0x60, 0x4, 0x36, 0x10, 0x60, 0x28, 0x57, 0x60, 0x0, 0x35, 0x60, 0xe0, 0x1c, 0x80, 0x63, 0xeb, 0x8a, 0xc9, 0x21, 0x14, 0x60, 0x2d, 0x57, 0x5b, 0x60, 0x0, 0x80, 0xfd, 0x5b, 0x60, 0x60, 0x60, 0x4, 0x80, 0x36, 0x3, 0x60, 0x40, 0x81, 0x10, 0x15, 0x60, 0x41, 0x57, 0x60, 0x0, 0x80, 0xfd, 0x5b, 0x81, 0x1, 0x90, 0x80, 0x80, 0x35, 0x90, 0x60, 0x20, 0x1, 0x90, 0x92, 0x91, 0x90, 0x80, 0x35, 0x90, 0x60, 0x20, 0x1, 0x90, 0x92, 0x91, 0x90, 0x50, 0x50, 0x50, 0x60, 0x62, 0x56, 0x5b, 0x0, 0x5b, 0x81, 0x60, 0x0, 0x81, 0x90, 0x55, 0x50, 0x80, 0x82, 0x7f, 0xf3, 0xca, 0x12, 0x4a, 0x69, 0x7b, 0xa0, 0x7e, 0x8c, 0x5e, 0x80, 0xbe, 0xbc, 0xfc, 0xc4, 0x89, 0x91, 0xfc, 0x16, 0xa6, 0x31, 0x70, 0xe8, 0xa9, 0x20, 0x6e, 0x30, 0x50, 0x89, 0x60, 0xd0, 0x3, 0x60, 0x40, 0x51, 0x60, 0x40, 0x51, 0x80, 0x91, 0x3, 0x90, 0xa3, 0x50, 0x50, 0x56, 0xfe, 0xa2, 0x65, 0x62, 0x7a, 0x7a, 0x72, 0x31, 0x58, 0x20, 0x1d, 0x94, 0xd2, 0x18, 0x7a, 0xaf, 0x3a, 0x67, 0x90, 0x52, 0x7b, 0x61, 0x5f, 0xcc, 0x40, 0x97, 0xf, 0xeb, 0xf0, 0x38, 0x5f, 0xa6, 0xd7, 0x2a, 0x23, 0x44, 0x84, 0x8e, 0xbd, 0xd, 0xf3, 0xe9, 0x64, 0x73, 0x6f, 0x6c, 0x63, 0x43, 0x0, 0x5, 0x11, 0x0, 0x32}
        	            	actual  : hexutil.Bytes(nil)
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,16 +1,2 @@
        	            	-(hexutil.Bytes) (len=208) {
        	            	- 00000000  60 80 60 40 52 34 80 15  60 0f 57 60 00 80 fd 5b  |`.`@R4..`.W`...[|
        	            	- 00000010  50 60 04 36 10 60 28 57  60 00 35 60 e0 1c 80 63  |P`.6.`(W`.5`...c|
        	            	- 00000020  eb 8a c9 21 14 60 2d 57  5b 60 00 80 fd 5b 60 60  |...!.`-W[`...[``|
        	            	- 00000030  60 04 80 36 03 60 40 81  10 15 60 41 57 60 00 80  |`..6.`@...`AW`..|
        	            	- 00000040  fd 5b 81 01 90 80 80 35  90 60 20 01 90 92 91 90  |.[.....5.` .....|
        	            	- 00000050  80 35 90 60 20 01 90 92  91 90 50 50 50 60 62 56  |.5.` .....PPP`bV|

```

**Review by noot — COMMENTED** [review]
the solidity tests fail on this branch with `Error: while migrating Migrations: The contract code couldn't be stored, please check your gas limit.`

same with deploying aragonOS. this is probably why the export account test is failing

I can't tell exactly why from looking at the changes, going to look into it more

**Comment by fedekunze** [comment]
> the solidity tests fail on this branch with Error: while migrating Migrations: The contract code couldn't be stored, please check your gas limit.
same with deploying aragonOS. this is probably why the export account test is failing

The `ChainConfig` has to do with the gas cost updates. Prob looking into that is the first thing we should do

**Review by noot — APPROVED** [review]
contract deployments work for me now, looks good!

File: tests/rpc_test.go [hunk 2]
```diff
@@ -754,11 +754,11 @@ func TestEth_EstimateGas(t *testing.T) {
 	require.NotNil(t, rpcRes)
 	require.NotEmpty(t, rpcRes.Result)
 
-	var gas hexutil.Bytes
+	var gas string
 	err := json.Unmarshal(rpcRes.Result, &gas)
 	require.NoError(t, err, string(rpcRes.Result))
 
-	require.Equal(t, "0xffdf", gas.String())
+	require.Equal(t, "0x1051d", gas)
```
Comment (fedekunze) [inline tests/rpc_test.go:761]: not sure if we should address this? technically it should remain the same? @noot

Comment (noot) [reply tests/rpc_test.go:761]: yeah it should likely be the same, do we have any tests for actual gas consumption vs estimation? if not we should add some

**Pull request merged**